File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_12.java
package org.twyszomirski.aoc.solutions;


import io.vavr.Tuple2;
import io.vavr.collection.Stream;
import org.twyszomirski.aoc.utils.Pair;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.IntStream;

import static java.io.IO.println;

public class Day_12 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_12.txt"));
        println("======== Day 12 ===========");
        part_1(lines);
        part_2(lines);
    }

    private static void part_1(List<String> lines) {
        var data = getData(lines);
        var processed = new HashSet<Plant>();
        var regions = data.stream().flatMap(Collection::stream)
                .filter(it -> !processed.contains(it))
                .map(it -> findRegion(data, it, processed));

        var sum = regions.map(region -> new Pair<>((long) region.size(),
                        region.stream().mapToLong(plant -> (long) countBorders(data, plant)).sum()))
                .mapToLong(metric -> metric.left() * metric.right()).sum();
        println(sum);
    }

    private static void part_2(List<String> lines) {
        var data = getData(lines);
        var processed = new HashSet<Plant>();
        var regions = data.stream().flatMap(Collection::stream)
                .filter(it -> !processed.contains(it))
                .map(it -> findRegion(data, it, processed));

        var regionBorders = regions.map(region -> new Tuple2<>(region, dropBordersOnGrid(data, region)));
        var metrics = regionBorders.map(regionAndBorders -> new Tuple2<>(regionAndBorders._1.size(), (scanHorizontal(regionAndBorders._2) + scanVertical(regionAndBorders._2))));
        var sum = metrics.mapToLong(m -> m._1 * m._2).sum();

        println(sum);
    }

    private static List<Plant> findRegion(List<List<Plant>> grid, Plant plant, Set<Plant> processed) {
        var left = getNeighbour(grid, plant, -1, 0);
        var right = getNeighbour(grid, plant, 1, 0);
        var up = getNeighbour(grid, plant, 0, -1);
        var down = getNeighbour(grid, plant, 0, 1);

        var result = new ArrayList<Plant>();
        processed.add(plant);
        result.add(plant);

        result.addAll(Stream.of(left, right, up, down).filter(Objects::nonNull)
                .filter(it -> Objects.equals(it.symbol, plant.symbol))
                .filter(it -> !processed.contains(it))
                .map(next -> {
                    processed.add(next);
                    return next;
                })
                .flatMap(next ->
                        Stream.ofAll(findRegion(grid, next, processed))

                ).toJavaList());

        return result;
    }

    private static Integer countBorders(List<List<Plant>> grid, Plant plant) {
        return getBorders(grid, plant).size();
    }

    private static List<List<List<Border>>> dropBordersOnGrid(List<List<Plant>> grid, List<Plant> region) {
        return IntStream.range(0, grid.size()).mapToObj(r -> {
                    return IntStream.range(0, grid.getFirst().size()).mapToObj(c -> {
                        var p = region.stream().filter(x -> x.row == r && x.column == c).findFirst();
                        return p.map(plant -> getBorders(grid, plant)).orElse(null);
                    }).toList();
                }
        ).toList();
    }

    private static Long scanHorizontal(List<List<List<Border>>> grid) {
        return Stream.of(Border.UP, Border.DOWN).map(border -> Stream.ofAll(grid)
                .map(line -> countBorders(Stream.ofAll(line).map(pos -> {
                    if (pos != null && pos.contains(border)) return "1";
                    else return "0";
                }).toJavaList())).sum().longValue()).sum().longValue();
    }

    private static Long scanVertical(List<List<List<Border>>> grid) {
        return Stream.of(Border.LEFT, Border.RIGHT)
                .map(border -> {
                            return IntStream.range(0, grid.getFirst().size()).mapToLong(col ->
                            {
                                var column = IntStream.range(0, grid.size()).mapToObj(r -> grid.get(r).get(col));
                                var a = countBorders(column.map(pos -> {
                                    if (pos != null && pos.contains(border)) return "1";
                                    else return "0";
                                }).toList());
                                return a;
                            }).sum();

                        }
                ).sum().longValue();
    }

    private static Long countBorders(List<String> line) {
        return Arrays.stream(String.join("", line).split("0")).filter(it -> it.contains("1")).count();
    }

    private static List<Border> getBorders(List<List<Plant>> grid, Plant plant) {
        var left = getNeighbour(grid, plant, -1, 0);
        var right = getNeighbour(grid, plant, 1, 0);
        var up = getNeighbour(grid, plant, 0, -1);
        var down = getNeighbour(grid, plant, 0, 1);

        var ver = Stream.of(
                        new Tuple2<>(left, Border.LEFT),
                        new Tuple2<>(right, Border.RIGHT)
                ).filter(it -> it._1 == null || !Objects.equals(it._1.symbol, plant.symbol))
                .map(Tuple2::_2).toJavaList();
        var hor = Stream.of(
                        new Tuple2<>(up, Border.UP),
                        new Tuple2<>(down, Border.DOWN)
                ).filter(it -> it._1 == null || !Objects.equals(it._1.symbol, plant.symbol))
                .map(Tuple2::_2).toJavaList();
        return Stream.ofAll(ver).appendAll(hor).toJavaList();
    }

    private static Plant getNeighbour(List<List<Plant>> grid, Plant point, int horizontal, int vertical) {
        var column = point.column + horizontal;
        var row = point.row + vertical;

        if (row >= 0 && column >= 0 && row < grid.size() && column < grid.getFirst().size()) {
            return grid.get(row).get(column);
        }
        return null;
    }

    private static List<List<Plant>> getData(List<String> lines) {
        return Stream.ofAll(lines).zipWithIndex().map(row ->
                Stream.ofAll(Arrays.asList(row._1.split("")))
                        .zipWithIndex().map(column ->
                                new Plant(row._2, column._2, column._1)).toJavaList()).toJavaList();

    }


    record Plant(Integer row, Integer column, String symbol) {
    }

    enum Border {
        UP,
        DOWN,
        RIGHT,
        LEFT
    }

}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_12.kt
package org.twyszomirski.aoc.solutions

import java.io.File
import java.util.concurrent.atomic.AtomicInteger


class Day_12 {

    fun solve() {
        val lines = File("src/main/resources/input_day_12.txt").readLines()
        println("======== Day 12 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val data = readData(input)

        val indexes = mutableMapOf<String, AtomicInteger>()

        data.flatten().forEach { parcel ->
            if (!parcel.marked) {
                val index = indexes.computeIfAbsent(parcel.symbol) { AtomicInteger(0) }
                markNeighbours(data, parcel, index.andIncrement)
            }
        }

        val regions = data.flatten().groupBy { it.symbol + it.index }

        val metrics = regions.map { (_, parcels) ->
            Pair(parcels.size, parcels.sumOf { countBorders(data, it) })
        }

        val result = metrics.sumOf { it.first * it.second }

        println(result)
    }

    fun part2(input: List<String>) {
        val data = readData(input)

        val indexes = mutableMapOf<String, AtomicInteger>()

        data.flatten().forEach { parcel ->
            if (!parcel.marked) {
                val index = indexes.computeIfAbsent(parcel.symbol) { AtomicInteger(0) }
                markNeighbours(data, parcel, index.andIncrement)
            }
        }

        data.flatten().forEach { markBorders(data, it) }

        val regions = data.flatten().groupBy { Pair(it.symbol, it.index) }

        val metrics = regions.map { (key, parcels) ->
            Pair(
                parcels.size, scanHorizontal(data, key.first, key.second) +
                        scanVertical(data, key.first, key.second)
            )
        }

        val result = metrics.sumOf { it.first * it.second }

        println(result)
    }


    fun markNeighbours(matrix: PlantMatrix, parcel: PlantParcel, index: Int) {
        parcel.marked = true
        parcel.index = index
        val right = getNeighbour(matrix, parcel, 1, 0)
        val left = getNeighbour(matrix, parcel, -1, 0)
        val bottom = getNeighbour(matrix, parcel, 0, 1)
        val up = getNeighbour(matrix, parcel, 0, -1)

        listOfNotNull(left, right, up, bottom).forEach {
            if (it.symbol == parcel.symbol && !it.marked) {
                it.marked
                it.index = index
                markNeighbours(matrix, it, index)
            }
        }
    }

    fun countBorders(matrix: PlantMatrix, parcel: PlantParcel): Int {
        val left = getNeighbour(matrix, parcel, -1, 0)
        val right = getNeighbour(matrix, parcel, 1, 0)
        val up = getNeighbour(matrix, parcel, 0, -1)
        val down = getNeighbour(matrix, parcel, 0, 1)

        return listOf(left, right, up, down).map { neighbour ->
            when {
                neighbour == null -> 1
                neighbour.symbol != parcel.symbol -> 1
                else -> 0
            }
        }.sum()
    }

    fun getNeighbour(matrix: PlantMatrix, parcel: PlantParcel, hor: Int, ver: Int): PlantParcel? {
        val row = parcel.row + ver
        val col = parcel.column + hor

        if (row >= 0 && col >= 0 && row < matrix.size && col < matrix.first().size) {
            return matrix[row][col]
        }
        return null
    }

    fun markBorders(matrix: PlantMatrix, parcel: PlantParcel) {
        val right = getNeighbour(matrix, parcel, 1, 0)
        val left = getNeighbour(matrix, parcel, -1, 0)
        val bottom = getNeighbour(matrix, parcel, 0, 1)
        val up = getNeighbour(matrix, parcel, 0, -1)

        parcel.ver = listOf(
            left to Border.LEFT,
            right to Border.RIGHT
        ).filter { it.first == null || it.first!!.symbol != parcel.symbol }.map { it.second }
        parcel.hor = listOf(
            up to Border.UP,
            bottom to Border.DOWN
        ).filter { it.first == null || it.first!!.symbol != parcel.symbol }.map { it.second }
    }

    fun numberOfBorders(line: List<PlantParcel>, symbol: String, index: Int, check: (PlantParcel) -> Boolean): Int {

        val mapped = line.map { if (it.symbol == symbol && it.index == index && check(it)) 1 else 0 }

        return mapped.joinToString("")
            .split("0").count { it.contains("1") }
    }

    fun scanHorizontal(matrix: PlantMatrix, symbol: String, index: Int): Int {
        return listOf(Border.UP, Border.DOWN).sumOf { border ->
            matrix.sumOf { row ->
                numberOfBorders(row, symbol, index) { it.hor.contains(border) }
            }
        }
    }

    fun scanVertical(matrix: PlantMatrix, symbol: String, index: Int): Int {
        return listOf(Border.LEFT, Border.RIGHT).sumOf { border ->
            (0..<matrix.first().size).map {
                matrix.map { r -> r[it] }
            }.sumOf { numberOfBorders(it, symbol, index) { it.ver.contains(border) } }
        }
    }

    fun readData(lines: List<String>): PlantMatrix {
        return lines.mapIndexed { row, line ->
            line.split("").filter { it.isNotBlank() }.mapIndexed { column, symbol ->
                PlantParcel(row, column, symbol)
            }.toList()
        }.toList()
    }

    fun print(matrix: PlantMatrix, how: (PlantParcel) -> String) {
        matrix.forEach {
            println(it.joinToString(" ") { how(it) })
        }
    }
}

data class PlantParcel(
    val row: Int,
    val column: Int,
    val symbol: String,
    var marked: Boolean = false,
    var index: Int = -1,
    var hor: List<Border> = emptyList(),
    var ver: List<Border> = emptyList()
)

enum class Border {
    UP,
    DOWN,
    LEFT,
    RIGHT
}

typealias PlantMatrix = List<List<PlantParcel>>
